### ISSUE TITLE
no-typos rule doesn't handle custom proptypes classes

### ISSUE THREAD

**Issue by modosc** [yannickcr/eslint-plugin-react#1389]
with the `7.2.1` -> `7.3.0` upgrade this rule started breaking on this code:
```
import * as MyPropTypes from 'lib/my-prop-types'

const Foo = ({ bar } = {} ) => <div className={bar} />

Foo.propTypes = {
  bar: MyPropTypes.bar.isRequired
}
```
with
```
  29:25  error  Typo in declared prop type: bar  react/no-typos
```
interestingly enough if i change to this:
```
Foo.propTypes = {
  bar: MyPropTypes.string.isRequired
}
```

then the rule passes, even though `MyPropTypes` doesn't export a `string` prop.

**Comment by jseminck** [comment]
Just for information, indeed we added logic to validate the specific PropTypes. Since `string` is a key that is exported from the official `prop-types` package, it doesn't fail. `bar` is not a prop type defined in the official `prop-types` package, so it fails.

Not sure how we should handle this case yet. Just explaining what changes and why the error is introduced. 😄

**Comment by ljharb** [comment]
This seems like a bug in no-typos - we should only be looking for typos on `React.PropTypes` or on an import/require of `prop-types`.

**Comment by jseminck** [comment]
I'm trying to work on this... I guess we need to support the following cases? Did I miss anything

```js
import PropTypes from "prop-types"
import { PropTypes } from "react";
import { PropTypes as ValidPropTypes } from "react"
import React from "react" + React.PropTypes
```

**Comment by ljharb** [comment]
@jseminck also `require('prop-types')` in any form.

**Comment by martin-svk** [comment]
It also fail when importing specific prop types directly:

```
MyComponent.propTypes = {
    numberProp: number.isRequired,
    funcProp: func.isRequired
};
```

Resulting in `react/no-typos` Error.

**Comment by oieduardorabelo** [comment]
same happens when using [`react-intl`](https://github.com/yahoo/react-intl) inject hoc:

```js
import { injectIntl, intlShape } from "react-intl";

function Example() { //code }

Example.propTypes = {
  // ESLint error: 'Typo in declared prop type: isRequired (react/no-typos)'
  intl: intlShape.isRequired
};

export default injectIntl(Example)
```

![screen shot 2017-10-10 at 9 56 00 am](https://user-images.githubusercontent.com/829902/31358227-3ea76112-ada1-11e7-9fae-b9357da61283.png)

*you can ignore the line number in the screenshot, is just an example*

**Comment by felpin** [comment]
Well, this does not resolve the issue, but you can [disable a rule in ESLint with a comment](https://eslint.org/docs/user-guide/configuring#disabling-rules-with-inline-comments)

```javascript
Foo.propTypes = {
  // eslint-disable-next-line react/no-typos
  bar: MyPropTypes.string.isRequired
}
```

In this particular case I do that while this issue is open

**Comment by neiker** [comment]
Same using `react-immutable-proptypes`

**Comment by zxlin** [comment]
@jseminck 

Another case where webpack's `ProvidePlugin` is used to avoid repetitively importing `prop-types` everywhere. 

In my code I don't import `prop-types` explicitly and simply write:
```
const { string, object } = PropTypes;
```
and get the validators that way as webpack will replace all instances of `PropTypes` with the `prop-types` module.

**Comment by ljharb** [comment]
@zxlin using webpack to get magic implicit imports isn't necessarily a use case we should be handling or encouraging, imo.

**Comment by the-noob** [comment]
Adding this for whoever looks for MobX no-typos error :)

```JavaScript
import { observer, inject, PropTypes as MobXPropTypes } from 'mobx-react';

class App extends React.Component { ...}

App.wrappedComponent.propTypes = {
  appStore: MobXPropTypes.objectOrObservableObject.isRequired,
};
```
Produces
> Typo in declared prop type: objectOrObservableObject  react/no-typos

**Comment by jseminck** [comment]
I'm working on this, so that custom PropTypes imports will be ignored and only the real "prop-types" package will be considered for checking typos.

**Comment by ljharb** [comment]
@jseminck also `React.PropTypes`, hopefully :-)

**Comment by Kenjirou** [comment]
Same typo-error for ReactNative's style proptypes:

```javascript
import PropTypes from 'prop-types';
import {View} from 'react-native';

CoolComponent.propTypes = {
    containerStyle: View.propTypes.style,
};
```

**Comment by abdennour** [comment]
@jseminck  , any update ? we are waiting ☕️

**Comment by Hypnosphi** [comment]
@abdennour you can make your own PR if you feel like it

**Comment by jseminck** [comment]
I try to push a branch soon but it's harder than I had thought :P too many cases to think of.

**Comment by extwiii** [comment]
any update ? we are waiting ☕️

**Comment by alex-shamshurin** [comment]
Yes, I have the same issue. It works inside React.component extended class and doesn't work for stateless components

**Comment by Koleok** [comment]
I just need to drop a note on the repeated toe-tapping scoldy comments in this thread

_ie:_
> _any update? we are waiting ☕️_

**Github automatically posts references for:**
- opens/closes/merges on related MRs
- commits with references to this issue
- references to this issue from other issues

These **are** the updates you are looking for.
![updates](http://i0.kym-cdn.com/entries/icons/facebook/000/018/682/obi-wan.jpg)

It is not, and _should not be_ the responsibility of anyone involved in an OSS project to look through each issue every day and copy/paste links to commits or give _progress reports_ to impatient/unmotivated consumers of the library.

The whole point of open source, is that maintainers do things when they can, and _contributors_ who have time step in and fill the gaps when they can.

If you find yourself continually posting things in github issues like
> _"hey i need this to work right now how long till you fix it!"_

then guess what, **you** my friend are ready to be an open source contributor. 

You might say:
> _"I don't know how this library works!"_

well, read the source, and soon you will

> _"i don't have time my job is relentless and I'm being pressured"_

well, so are the maintainers you are pressuring, probably more so!

> _"No one has committed anything in a long time and i fear the project is abandoned :scream:"_

well, fork it and do it your way, or just be patient :)

**Comment by dkrutsko** [comment]
Hey everyone, I'd like to add two additional pieces of information about this issue:

1. Since PR #1652 it seems that `no-typos` started detecting more issues! However, While this rule detects "invalid" propTypes defined in functional components, it doesn't work inside classes when defining propTypes through static instance variables.

2. For this issue specifically, I'm guessing that all that needs to change is [this variable](https://github.com/yannickcr/eslint-plugin-react/blob/f4cab9a3ddeb730bf5f0ae47b739a6bba84ac237/lib/rules/no-typos.js#L25)? Would mutating it through props solve the issue?

**Comment by brettdh** [comment]
@jseminck how critical is it to have the behavior in the failing test? Your branch resolves my issue perfectly; I bet that's the case for a lot of folks here. The "renamed React import" seems like a very small corner case; I'd suggest breaking off a new issue for just that, and focus on getting the awesome work you've already done merged.

Since I'm invested in this now, I'll remove that test and create a PR. If this isn't acceptable, I'm sure the reasons will be discussed there. :)

@dkrutsko could you be a little more specific about what you think the fix is? Are you referring to the failing test @jseminck mentioned above, or are you suggesting that the fix for this entire issue is somehow much simpler than @jseminck 's changes?

If necessary, I'm happy to add your suggestions to my forthcoming PR.

**Comment by dkrutsko** [comment]
@brettdh It appears that the "valid" prop-types are retrieved using `Object.keys (require ('prop-types'))`. I'm guessing that making this variable user editable through the linter settings would allow anybody to add additional prop-types. Like in my case, I'd like an option to add more prop-types on top of the existing ones in the package.

**Comment by brettdh** [comment]
Ah, yep - that makes sense. I think I was confused by the overloaded term "prop" in your comment. :)

I think that's also a separate issue from this one. That is, I see two separable desired behaviors:
1. Correctly specified custom proptypes should not cause `no-typos` errors (this issue)
1. Custom proptypes with typos *should* cause `no-typos` errors (your request)

I can see why the second is valuable, but the first strikes me as higher-impact. Thus, I'd suggest you make a new issue to track just the second, as it seems ever so slightly out of scope for this one.

Sound reasonable?

**Comment by wpwood** [comment]
I think i may have another case where this is failing. When the shape is defined in the same file where it's used, it's failing. I have a file with two components, sharing a shape. Something like this:
```
const linkShape = PropTypes.shape({
  text: PropTypes.string,
  url: PropTypes.string,
})

FooterLinks.Link.propTypes = {
  link: linkShape.isRequired,
}
```
This rule gives the `Typo in declared prop type: isRequired  react/no-typos` error.
